gammapy/background/reflected.py

```diff
@@ -1,68 +1,19 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 import logging
 import numpy as np
-from scipy.ndimage import distance_transform_edt
-from astropy.coordinates import Angle
-from regions import PixCoord, CirclePixelRegion
-from ..maps import WcsNDMap
+from astropy.coordinates import Angle, SkyCoord
+from astropy import units as u
+from regions import PixCoord, CirclePixelRegion, CircleAnnulusPixelRegion, \
+                     RectanglePixelRegion, RectangleAnnulusPixelRegion, \
+                     EllipsePixelRegion,EllipseAnnulusPixelRegion
+from ..maps import WcsNDMap, WcsGeom
 from .background_estimate import BackgroundEstimate
 
 __all__ = ["ReflectedRegionsFinder", "ReflectedRegionsBackgroundEstimator"]
 
 log = logging.getLogger(__name__)
 
 
-def _compute_distance_image(mask_map):
```

Comment:
This PR removes the distance image implementation. It is used to make reflected region finding fast for the common case of circular regions. I think this speed-up really is needed: one of my first contributions to the HESS software was to add this, after running spectral analyses for HGPS sources. What I found was that reflected region computation was dominating analysis time overall completely - even event reconstruction, gamma-hadron separation, spectral likelihood fitting, everything else was much faster.

Can you please put it back and keep reflected region analysis fast for the common case of having circles (like for AGN)?

Reply:
I will make comparison of speed tests, put the results here, which will justify the retained solution.

Reply:
Your method using the optimised function '_compute_distance_image' is based on the assumption that a region has a radius (cf. _is_inside_exclusion). If one used an outer radius, more pixels than in reality will be considered as excluded, which is not optimal in this regards. In this PR, one uses the function 'contains' of a region, that garanties to have exact pixels.
Do you have an idea to adapt the function '_is_inside_exclusion' to be more accurate?

Reply:
The distance image algorithm is only for the case of a circular region.

Reply:
Ok. So, should I keep these algo only for the case of circular regions?
The pros: most of the case, we will have circular region, and so we gain some computing time.
The cons: more lines for just one case.
Any preference?

Reply:
Yes, please keep the fast case for circular regions.

Reply:
I have re-implemented this algorithm. The timing measurement are on the notebook intro_reflected_background using a circular region:
Christoph algo: total CPUTime=25.4s, Wall time=7.75s
Current algo using PixRegion.contains: total CPUTime=25.2s, Wall time=8.07s
The time different is tiny!
Should we thus keep an implementation using a different method depending on the On region type? My personal feeling is thus no.

Reply:
@bkhelifi - I don't see a notebook `intro_reflected_background`. All I see is https://github.com/gammapy/gammapy/blob/22eda59cf6feff6a29b421afc2a30a0158ec86f2/tutorials/intro_reflected_background.ipynb but it has no timings. Please share a link to the notebook.

A good way to share a notebook or some other file that contains something that's not supposed to be merged with a PR is on the side, e.g. in a repo in your account, or I often post a gist: https://gist.github.com/ .

Not sure what test case you measured. Here's one that could be used: https://docs.gammapy.org/0.11/background/reflected.html
The time will depend on the number of off regions tested, so a smaller on region (e.g. 0.1 deg) and larger FOV offset (e.g. 2 deg) so that there are ~ 100 off regions would make a good test case.
With the current algorithm in Gammapy master it will be fast, with that removed I would expect it to be slow.

Reply:
By using the code given in the web page that you mentioned, I made a fast notebook. Here are the results, first using your method, then the new one.

For each of the finder given in the notebook, the results for your method are:
CPU times: user 12.6 s, sys: 400 ms, total: 13 s; Wall time: 4.37 s
CPU times: user 12.1 s, sys: 291 ms, total: 12.4 s; Wall time: 4.04 s
CPU times: user 12.5 s, sys: 307 ms, total: 12.8 s; Wall time: 4.13 s

With the new method:
CPU times: user 12 s, sys: 413 ms, total: 12.4 s; Wall time: 4.32 s
CPU times: user 11.8 s, sys: 385 ms, total: 12.2 s; Wall time: 4.15 s
CPU times: user 11.9 s, sys: 356 ms, total: 12.3 s; Wall time: 4.18 s

When using an offset of 2deg, 23 reflected regions are built (r=0.3deg). The execution times are the the first finder:
Your method: CPU times: user 16 s, sys: 941 ms, total: 16.9 s; Wall time: 6.95 s
The new method: CPU times: user 15.5 s, sys: 837 ms, total: 16.3 s; Wall time: 7.3 s

The conclusions are the identical. The current algorithm is almost as fast as the previous one

Reply:
@bkhelifi - Can you please share the notebook in a gist so that I can try it out as well?

I'll be at a meeting the next days, will only be able to look at this again Monday.

@registerrier - If you have time to review this, please go ahead. If you're happy with the PR, feel free to merge, no need to wait for me.